### PR TITLE
Friendship v2

### PR DESCRIPTION
**Pull request by sergioobolevich**
In this milestone we did the following:
- Change the friendship implementation so we use 2 rows on the Friendship table instead of just one.
- User can see the friend requests and accept or reject them.
- Change the confirm(friend) method, so we create the second row in the Friendship table at that moment.

### CODE REVIEW

**Review by acushlakoncept — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️

Great job! thus far 👏  You have done well on this project.
Your specs are looking good 👍 
However, there are still some changes that are necessary for this project to be approved.

- [x] Having worked on your project and reached this milestone, by now your readme file should begin to look a bit more professional, kindly update it to reflect what you've done so far and remove all the TBA and make them available 😄 

- [x] In this milestone, we are required to implement friendships with 2 rows for mutual friendship. When friendship is confirmed, an additional row with reversed user-friend pair is created.
```ruby
# Friendship model

def confirm_friend
  self.update_attributes(confirmed: true)
  Friendship.create!(friend_id: self.user_id,
                  user_id: self.friend_id,
                  confirmed: true)
end
```

- [x] **Thanks to this, we can find user friends only by user_id in the friendships table**

```ruby
# User model

has_many :confirmed_friendships, -> { where confirmed: true }, class_name: "Friendship"
has_many :friends, through: :confirmed_friendships
```

 - [x] User#pending_friends can be defined as association with SQL query instead of processing objects in application memory (as the database was created to store and query data it is always more efficient than ruby)

```ruby
# User model

# Users who needs to confirm friendship
has_many :pending_friendships, -> { where confirmed: false }, class_name: "Friendship", foreign_key: "user_id"
has_many :pending_friends, through: :pending_friendships, source: :friend
```
 - [x] User#friend_requests can also be defined as association with SQL query instead of processing objects in application memory (as the database was created to store and query data it is always more efficient than ruby)

```ruby
# User model

# Users who requested to be friends (needed for notifications)
has_many :inverted_friendships, -> { where confirmed: false }, class_name: "Friendship", foreign_key: "friend_id"
has_many :friend_requests, through: :inverted_friendships
```

 - [x] In order to get all posts written by a user on their friends, you can use a simple method in User model (in order to NOT do it with Enumerables)

```ruby
# User model

def friends_and_own_posts
  Post.where(user: (self.friends << self))
  # This will produce SQL query with IN. Something like: select * from posts where user_id IN (1,45,874,43);
end
```

* [x] There should be absolutely NO logic in views. Views are only for displaying information that became available to it from its corresponding controller not generating it. **Resources**
* [Where Do I Put My Code?](http://codefol.io/posts/Where-Do-I-Put-My-Code)
* [The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)
* [Rails View Helpers Best Practices](https://medium.com/@acushla4real/rails-view-helpers-best-practices-guide-d7d7b6db3db4)
* [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright

- [x] I see a lot of application logic as well in your views: There should be NO application logic in the views nor helpers (no Post.new, Comment.create, current_user.friends.find_by(...)...). Everything connected with data processing should be encapsulated in models and used in controllers.
Please do take a look at these view files and refactor them accordingly to do away with the logics in them, use the resources I have provided above to guide you through.
- All files inside `views/users` except `index.html.erb`
- `views/layouts/application.html.erb` file

#### kindly make the appropriate changes and ask for another code review 👍

### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla), if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

File: app/models/user.rb [hunk 1]
```diff
@@ -11,27 +11,37 @@ class User < ApplicationRecord
...
+  has_many :inverted_friendships, -> { where confirmed: nil }, class_name: 'Friendship', foreign_key: 'friend_id'
+  has_many :friend_requests, through: :inverted_friendships, source: :user
+  # def friends
+  #   friends_array = friendships.map { |friendship| friendship.friend if friendship.confirmed }
+  #   friends_array += inverse_friendships.map { |friendship| friendship.user if friendship.confirmed }
+  #   friends_array.compact
+  # end
+
+  # def pending_friends
+  #   friendships.map { |friendship| friendship.friend unless friendship.confirmed }.compact
```
Comment (voscarmv) [inline app/models/user.rb]: Please remove all the commented code in this and other files.

**Review by voscarmv — CHANGES_REQUESTED** [review]
# Impressive work, team. 💻 👀 ✨ 

- Excellent work applying all the changes requested by the previous TSE.

## However, there are some changes required 👀 🧰 🔨 🔧 

**[MANDATORY]**

- Just one last detail, from my point of view: please remove all commented code. You've left a lot of commented code for example in the `users` model file. Please remove this and any other commented code in your project.

### Submit these changes and request a new code review! 😄 👍 

### Questions or comments? Reach me on [Twitter](https://twitter.com/voscarmv) or [LinkedIn](https://linkedin.com/in/oscar-mier) 🧑‍💻🗨️🧑‍💻 
# Happy Hacking! 🧠 ⌨️ 🖱️ 🖥️

File: app/models/user.rb (hunk 1 above)
Comment (sergioobolevich) [reply app/models/user.rb]: Done, thanks for advicing!

**Review by SpaYco — APPROVED** [review]
# Project Approved ✔️ 

thank you for making the required changes, your project is now approved!

### please don't forget to merge to master, good luck with the next projects!


Best, 
Aziz Mejri 
_contact me on [twitter](https://twitter.com/iSpaYco) or slack (@SpaYco) if you need any help_

**Pull request merged**